File: Sources/MiniFoundationCompat/URL.swift
// SPDX-FileCopyrightText: 2026 Davide De Rosa
//
// SPDX-License-Identifier: MIT

internal import _MiniFoundationCore_C

extension Compat {
    public final class URL: MiniURLProtocol, Hashable, Codable, @unchecked Sendable, CustomStringConvertible {
        private let impl: OpaquePointer

        public init?(string: String) {
            guard let impl = minif_url_create(string) else {
                return nil
            }
            self.impl = impl
        }

        private init(impl: OpaquePointer) {
            self.impl = impl
        }

        deinit {
            minif_url_free(impl)
        }

        public var scheme: String? {
            var count = 0
            guard let str = minif_url_get_scheme(impl, &count) else { return nil }
            return str.sizedString(count: count)
        }

        public var host: String? {
            var count = 0
            guard let str = minif_url_get_host(impl, &count) else { return nil }
            let host = str.sizedString(count: count)
            guard !host.isEmpty else { return nil }
            return host
        }

        public var port: Int? {
            let port = minif_url_get_port(impl)
            return port > 0 ? Int(port) : nil
        }

        public var path: String {
            var count = 0
            var decodedCount = 0
            guard let str = minif_url_get_path(impl, &count) else { return "" }
            let decoded = minif_url_alloc_decoded(str, count, &decodedCount)
            let decodedString = UnsafePointer(decoded).sizedString(count: decodedCount)
            free(decoded)
            return decodedString
        }

        public var lastPathComponent: String {
            var count = 0
            guard let str = minif_url_get_last_path_component(impl, &count) else { return "" }
            return str.sizedString(count: count)
        }

        public var query: String? {
            var count = 0
            guard let str = minif_url_get_query(impl, &count) else { return nil }
            return str.sizedString(count: count)
        }

        public var fragment: String? {
            var count = 0
            guard let str = minif_url_get_fragment(impl, &count) else { return nil }
            return str.sizedString(count: count)
        }

        public var absoluteString: String {
            String(cString: minif_url_get_string(impl))
        }

        // MARK: CustomStringConvertible

        public var description: String {
            absoluteString
        }

        // MARK: Codable

        public init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            let string = try container.decode(String.self)
            guard let impl = minif_url_create(string) else {
                throw MiniFoundationError.decoding
            }
            self.impl = impl
        }

        public func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            try container.encode(absoluteString)
        }
    }
}

extension Compat.URL {
    public static func == (lhs: Compat.URL, rhs: Compat.URL) -> Bool {
        lhs.absoluteString == rhs.absoluteString
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(absoluteString)
    }
}

// MARK: File URLs

// FIXME: #303, Optimistic, implement and test everything here, esp. on Windows

extension Compat.URL {
    public convenience init(fileURLWithPath path: String) {
        let forbidden = "#?"
        guard path.rangeOfCharacter(from: Compat.CharacterSet(charactersIn: forbidden)) == nil else {
            fatalError("Path contains forbidden characters (\(forbidden)): \(path)")
        }
        // FIXME: #303, Does not handle Windows drive unit and backslashes
        let absPath: String
        if path.hasPrefix(String.pathSeparator) {
            absPath = path
        } else {
            let cwd = minif_os_alloc_current_dir()
            absPath = String(cString: cwd) + "/" + path
            free(UnsafeMutableRawPointer(mutating: cwd));
        }
        let urlString = "file://\(absPath)"
        self.init(string: urlString)!
    }

    public func filePath() -> String {
        path
    }

    public func miniAppending(component: String) -> Self {
        let newPath = "\(filePath())\(String.pathSeparator).\(component)"
        return Self(string: newPath)!
    }

    public func miniAppending(path: String) -> Self {
        let newPath = "\(filePath())\(String.pathSeparator).\(path)"
        return Self(string: newPath)!
    }

    public func miniAppending(pathExtension: String) -> Self {
        let newPath = "\(filePath()).\(pathExtension)"
        return Self(string: newPath)!
    }

    public func miniDeletingLastPathComponent() -> Self {
        var pathComponents = filePath().components(separatedBy: String.pathSeparator)
        pathComponents.removeLast()
        return Self(string: pathComponents.joined(separator: String.pathSeparator))!
    }
}

extension String {
    static var pathSeparator: String {
#if os(Windows)
        "\\"
#else
        "/"
#endif
    }

    func appendingPathComponent(_ component: String) -> String {
        "\(self)\(Self.pathSeparator)\(component)"
    }

    func deletingLastPathComponent() -> String {
        var comps = components(separatedBy: Self.pathSeparator)
        comps.removeLast()
        return comps.joined(separator: Self.pathSeparator)
    }
}

private extension UnsafePointer where Pointee == CChar {
    func sizedString(count: Int) -> String {
        withMemoryRebound(to: UInt8.self, capacity: count) {
            let buf = UnsafeBufferPointer(start: $0, count: count)
            return String(decoding: buf, as: UTF8.self)
        }
    }
}

File: Sources/MiniFoundationCore_C/url.c
/*
 * SPDX-FileCopyrightText: 2026 Davide De Rosa
 *
 * SPDX-License-Identifier: MIT
 */

#include "mini_foundation.h"
#include <stdlib.h>
#include <assert.h>
#include "url.h"

#define URL_SEPARATOR '/'

const char *memrchr(const char *s, int c, size_t n);

struct _minif_url {
    const char *original;
    const char *subject;
    URL impl;
};

minif_url *minif_url_create(const char *string) {
    URL impl = { 0 };
    char *original = minif_strdup(string);
    char *subject = minif_strdup(string);
    if (!original || !subject) goto failure;
    int len = (int)strlen(subject);
    if (len > 1) {
        const char *trailing = subject + len - 1;
        while (trailing != subject && *trailing == URL_SEPARATOR) {
            --len;
            --trailing;
        }
    }
    if (url_parse(subject, len, NULL, &impl, URL_FLAG_RFC3986) < 0) goto failure;
    minif_url *url = (minif_url *)calloc(1, sizeof(*url));
    url->original = original;
    url->subject = subject;
    url->impl = impl;
    return url;
failure:
    if (original) free(original);
    if (subject) free(subject);
    return NULL;
}

void minif_url_free(minif_url *url) {
    if (!url) return;
    free((void *)url->original);
    free((void *)url->subject);
    free(url);
}

const char *minif_url_get_string(minif_url *url) {
    return url->original;
}

const char *minif_url_get_scheme(minif_url *url, size_t *len) {
    if (url->impl.scheme.len == 0) return NULL;
    *len = url->impl.scheme.len;
    return url->impl.scheme.ptr;
}

const char *minif_url_get_host(minif_url *url, size_t *len) {
    if (url->impl.host_text.len == 0) return NULL;
    if (url->impl.host_type == URL_HOST_IPV6) {
        *len = url->impl.host_text.len - 2;
        return url->impl.host_text.ptr + 1;
    }
    *len = url->impl.host_text.len;
    return url->impl.host_text.ptr;
}

int minif_url_get_port(minif_url *url) {
    return url->impl.port;
}

const char *minif_url_get_path(minif_url *url, size_t *len) {
    if (url->impl.path.len == 0) return NULL;
    *len = url->impl.path.len;
    return url->impl.path.ptr;
}

const char *minif_url_get_query(minif_url *url, size_t *len) {
    if (url->impl.query.len == 0) return NULL;
    *len = url->impl.query.len;
    return url->impl.query.ptr;
}

const char *minif_url_get_fragment(minif_url *url, size_t *len) {
    if (url->impl.fragment.len == 0) return NULL;
    *len = url->impl.fragment.len;
    return url->impl.fragment.ptr;
}

const char *minif_url_get_last_path_component(minif_url *url, size_t *len) {
    if (url->impl.path.len == 0) return NULL;
    const char *p = memrchr(url->impl.path.ptr, URL_SEPARATOR, url->impl.path.len);
    // Return the full path
    if (!p) {
        *len = url->impl.path.len;
        return url->impl.path.ptr;
    }
    // Return the path after the URL_SEPARATOR
    ++p;
    const size_t offset = p - url->impl.path.ptr;
    *len = url->impl.path.len - offset;
    return p;
}

char *minif_url_alloc_decoded(const char *str, size_t len, size_t *dec_len) {
    // Force unsafe cast as url_percent_decode() has no side-effect
    URL_String comp = { (char *)str, (int)len };
    *dec_len = url_percent_decode(comp, NULL, 0);
    char *dec_str = (char *)calloc(1, *dec_len);
    const size_t redec_len = url_percent_decode(comp, dec_str, (int)*dec_len);
    assert(redec_len == *dec_len);
    return dec_str;
}

const char *memrchr(const char *s, int c, size_t n) {
    const char *p = (const char *)s + n;
    while (n--) {
        if (*--p == (char)c) return p;
    }
    return NULL;
}
